packages/middleware/rate-limiter/src/sliding-window.ts

```diff
@@ -0,0 +1,125 @@
+import { Rate, RateLiteral, parseRate } from './rate'
+import { Slot, SlotProvider } from './slot'
+import { types } from '@mondrian-framework/model'
+import { functions } from '@mondrian-framework/module'
+
+//inspired by https://blog.cloudflare.com/counting-things-a-lot-of-different-things/
+
+/**
+ * Rate limit at `rateLimit` in the `samplingPeriod`.
+ * Example of maximum 10 request every 2 minutes:
+ * ```typescript
+ * new SlidingWindow({ rateLimit: 10, samplingPeriod: "2 minutes" })
+ * ```
+ */
+export class SlidingWindowProvider {
+  private readonly slidingWindows: Map<string, SlidingWindow> = new Map()
+  private readonly rate: Rate
+  private readonly slotProvider: SlotProvider
+
+  constructor({ rate, slotProvider }: { rate: Rate | RateLiteral; slotProvider: SlotProvider }) {
+    this.rate = typeof rate === 'string' ? parseRate(rate) : rate
+    this.slotProvider = slotProvider
+  }
+
+  get({ key }: { key: string }): SlidingWindow {
+    const slidingWindow = this.slidingWindows.get(key)
+    if (slidingWindow) {
+      return slidingWindow
+    }
+    const newSlidingWindow = new SlidingWindow({ rate: this.rate, slotProvider: this.slotProvider, key })
+    this.slidingWindows.set(key, newSlidingWindow)
+    return newSlidingWindow
```

Comment:
This is a matter of personal preference but I'd use the explicit `else` branch, it also makes it easier to see the flow of the code and refactor things in the future

```suggestion
    if (slidingWindow) {
      return slidingWindow
    } else {
      const newSlidingWindow = new SlidingWindow({ rate: this.rate, slotProvider: this.slotProvider, key })
      this.slidingWindows.set(key, newSlidingWindow)
      return newSlidingWindow
    }
```